source/client.c

```diff
@@ -94,11 +99,16 @@ static void s_request_timeout(struct aws_channel_task *channel_task, void *arg,
     struct aws_mqtt_client_connection *connection = timeout_task_arg->connection;
 
     if (status == AWS_TASK_STATUS_RUN_READY) {
-        if (!timeout_task_arg->cancelled) {
+        if (timeout_task_arg->task_arg_wrapper != NULL) {
             mqtt_request_complete(connection, AWS_ERROR_MQTT_TIMEOUT, timeout_task_arg->packet_id);
         }
     }
 
+    if (timeout_task_arg->task_arg_wrapper != NULL) {
+        timeout_task_arg->task_arg_wrapper->timeout_task_arg = NULL;
```

Comment:
`/* NULL out pointers so that, when completion function runs, it knows that the timeout task has already been freed */